### ISSUE TITLE
Wrong reference on list of injected beans

### ISSUE THREAD

**Issue by dcdh** [quarkusio/quarkus#39310]
### Describe the bug

I am producing multiple synthetic beans this way:

```java
    @BuildStep
    @Record(ExecutionTime.STATIC_INIT)
    List<SyntheticBeanBuildItem> registerJdbcLockProviderInitializer(
            final ApplicationIndexBuildItem applicationIndexBuildItem,
            final DataSourceNameRecorder dataSourceNameRecorder) {
        final List<String> dataSourceNames = getDataSourcesNameFromJdbcSchedulerLocks(applicationIndexBuildItem);
        List<SyntheticBeanBuildItem> list = dataSourceNames.stream()
                .map(dataSourceName -> SyntheticBeanBuildItem.configure(DataSourceName.class)
                        .scope(Singleton.class)
                        .identifier(dataSourceName)
                        .supplier(dataSourceNameRecorder.dataSourceNameSupplier(dataSourceName))
                        .unremovable()
                        .done())
                .toList();
        return list;
    }
```

I am using a supplier to pass the argument and create the bean but it also does not worked when using createWith this way

```java
    @BuildStep
    @Record(ExecutionTime.STATIC_INIT)
    List<SyntheticBeanBuildItem> registerJdbcLockProviderInitializer(
            final ApplicationIndexBuildItem applicationIndexBuildItem,
            final DataSourceNameRecorder dataSourceNameRecorder) {
        final List<String> dataSourceNames = getDataSourcesNameFromJdbcSchedulerLocks(applicationIndexBuildItem);
        return dataSourceNames.stream()
                .map(dataSourceName -> SyntheticBeanBuildItem.configure(DataSourceName.class)
                        .scope(Singleton.class)
                        .identifier(dataSourceName)
                        .createWith(dataSourceNameRecorder.dataSourceNameFunction(dataSourceName))
                        //                        .supplier(dataSourceNameRecorder.dataSourceNameSupplier(dataSourceName))
                        .unremovable()
                        .done())
                .toList();
    }
```

```java
@Recorder
public class DataSourceNameRecorder {
    public Supplier<DataSourceName> dataSourceNameSupplier(final String dataSourceName) {
        return () -> (DataSourceName) () -> dataSourceName;
    }

    public Function<SyntheticCreationalContext<DataSourceName>, DataSourceName> dataSourceNameFunction(
            final String dataSourceName) {
        return dataSourceNameSyntheticCreationalContext -> (DataSourceName) () -> dataSourceName;
    }
}
```

and all the beans are initialized using the expected value but when it will call my method to return the dataSourceName, the last value is returned.

I guess there is a reference error in my injected list of DataSourceName here:
```java
  private final List<DataSourceName> dataSourcesName;
```

### Expected behavior

When running the integration test on  **integration-tests** / **jdbc** I should have two `DataSourceName` beans. One having `<default>` as value and another one having `master` as value.

### Actual behavior

Currently I have got two beans having the same value `master`.

### How to Reproduce?

1. git clone https://github.com/dcdh/quarkus-shedlock
2. Checkout the release/1.0.0 branch
3. Put a debug endpoint on `JdbcLockProviderInitializer`
4. Run in debug the integration jdbc
5. You will notice that two beans `dataSourcesName` are injected and have same value while in processor the supplier are called twice with different data source name.

### Output of `uname -a` or `ver`

Linux 2a02-8428-dff8-c601-234b-8c10-a3c4-2308.rev.sfr.net 6.6.11-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 10 19:25:59 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.9" 2023-10-17

### Quarkus version or git rev

3.8.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.1 (Red Hat 3.9.1-3)

**Comment by Ladicek** [comment]
I see where the problem is: the map from which a creation function (or supplier) is looked up from at runtime has incorrectly generated keys, see https://github.com/quarkusio/quarkus/blob/3.8.2/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeansProcessor.java#L95-L98

The fix is to make [at least] the bean identifier part of the key.

**Comment by Ladicek** [comment]
Should be fixed by #39402
